### PR TITLE
Feature/e2e test search

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

Backend and API end to end test of sunny day scenario.

### How to review

Run all services (whilst on cmd-develop) make sure you have configured all API databases to have values of`test`. Then cd into endToEndTests and run `go test ./...`; make sure test passes.

Make sure all data is cleaned up after test has run, hence all collections are empty under test database in mongo db, neo4j is clean of test data and elasticsearch (es) has no test indices (you can use `curl -XGET 'localhost:9200/_aliases?pretty'` for es).

Check test logic and that any comments make sense.

### Who can review

Team B

### CODE REVIEW

File: endToEndTests/end_to_end_api_test.go [hunk 1]
```diff
@@ -113,46 +116,48 @@ func TestSuccessfulEndToEndProcess(t *testing.T) {
...
 				log.ErrorC("Timed out - failed to get instance document to a state of completed", err, log.Data{"instance_id": instanceID, "state": instanceResource.State, "timeout": timeout})
 				os.Exit(1)
 			}
 
-			So(instanceResource.Headers, ShouldResemble, &[]string{"V4_0", "Time_codelist", "Time", "Geography_codelist", "Geography", "cpi1dim1aggid", "Aggregate"})
-			So(instanceResource.InsertedObservations, ShouldResemble, &totalObservations)
-			So(instanceResource.State, ShouldEqual, "completed")
-			So(instanceResource.TotalObservations, ShouldResemble, &totalObservations)
+			So(instanceResource.Headers, ShouldResemble, &[]string{"V4_0", "time", "time", "uk-only", "geography", "cpih1dim1aggid", "aggregate"})
+			//So(instanceResource.ImportTasks.ImportObservations.InsertedObservations, ShouldResemble, totalObservations)
```
Comment (Matt-Guest) [inline endToEndTests/end_to_end_api_test.go]: There a commented out assert here

File: endToEndTests/initialise.go [hunk 2]
```diff
@@ -136,3 +144,17 @@ func deleteMongoTestData(datasetID string) bool {
...
+
+func generateGenericHierarchy() error {
+	datastore, err := neo4j.NewDatastore(cfg.Neo4jAddr, "", neo4j.GenericHierarchyCPIHTestData)
+	if err != nil {
+		log.ErrorC("Unable to connect to neo4j", err, nil)
+		return err
+	}
+
+	if err = datastore.CreateGenericHierarchy("cpih1dim1aggid"); err != nil {
+		return err
```
Comment (Matt-Guest) [inline endToEndTests/initialise.go]: The error should be logged here

File: searchAPI/setup.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/ONSdigital/dp-api-tests/testDataSetup/elasticsearch"
+	"github.com/ian-kent/go-log/log"
+)
+
+func createSearchIndex(url, instanceID, dimension string) error {
+	log.Info("nath got here", nil)
```
Comment (Matt-Guest) [inline searchAPI/setup.go]: Can the log message be changed here

File: testDataSetup/elasticsearch/elasticsearch.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+// CreateSearchIndex represents the creation and loading of test data into an index for testing
+func (i *Index) CreateSearchIndex() error {
+	index := i.URL + "/" + i.InstanceID + "_" + i.Dimension
+	// Remove index
+	statusCode, err := DeleteIndex(index)
+	if err != nil {
+		if statusCode != http.StatusNotFound {
+			log.ErrorC("failed to delete index", err, log.Data{"path": index})
+			return err
```
Comment (Matt-Guest) [inline testDataSetup/elasticsearch/elasticsearch.go:82]: Should this trigger `os.Exit(1)`? So its like the mongodb setup

File: testDataSetup/elasticsearch/elasticsearch.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	_, _, err = CallElastic(ctx, index, "PUT", indexMappings)
+	if err != nil {
+		log.ErrorC("fail to create index", err, log.Data{"path": index})
+		return err
+	}
+
+	// Add docs to index
+	file, err := os.Open(i.TestDataFile)
+	if err != nil {
+		return err
```
Comment (Matt-Guest) [inline testDataSetup/elasticsearch/elasticsearch.go]: Can the error be logged?

File: testDataSetup/elasticsearch/elasticsearch.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+			return err
+		}
+	}
+
+	ctx := context.Background()
+
+	// Create index
+	indexMappings, err := ioutil.ReadFile(i.MappingsFile)
+	if err != nil {
+		return err
```
Comment (Matt-Guest) [inline testDataSetup/elasticsearch/elasticsearch.go]: can the error be logged?

File: testDataSetup/neo4j/neo4j.go [hunk 7]
```diff
@@ -95,3 +98,35 @@ func (ds *Datastore) Setup() error {
...
 }
+
+// CreateGenericHierarchy the neo4j database
+func (ds *Datastore) CreateGenericHierarchy(hierarchyCode string) error {
+	_, err := ds.connection.ExecNeo("MATCH (n:`_generic_hierarchy_node_"+hierarchyCode+"`) DETACH DELETE n", nil)
+	if err != nil {
+		return err
+	}
+
+	//b, err := ioutil.ReadFile(ds.testData)
```
Comment (Matt-Guest) [inline testDataSetup/neo4j/neo4j.go]: commented out code

File: searchAPI/setup.go (hunk 3 above)
Comment (nshumoogum) [reply searchAPI/setup.go]: whoops, been on holiday! ;p

File: testDataSetup/elasticsearch/elasticsearch.go (hunk 4 above)
Comment (nshumoogum) [reply testDataSetup/elasticsearch/elasticsearch.go:82]: No the thing calling this method would need to handle the error

**Review by Matt-Guest — APPROVED** [review]

**Pull request merged**